lv_timer_handler hangs when running on separate thread on ESP32 using FREERTOS using Arduino framework

### LVGL version

9.1.0

### What happened?

When I call lv_timer_handler from the main loop (with or without a mutex protection everything works.

When I create a separate thread using xTaskCreatePinnedToCore (on either core) with or without a mutex protection and this task calls the method lv_timer_handler this method never returns.

### How to reproduce?

//This is the task definition:
void ProcessDisplay(void *arg)
{
    ESP_LOGD(TAG, "Starting LVGL task");

    uint32_t task_delay_ms = LVGL_PORT_TASK_MAX_DELAY_MS;
    while (1) {
        lv_mutex_lock(&lvgl_mutex);
        task_delay_ms = lv_timer_handler(); 
        lv_mutex_unlock(&lvgl_mutex);
        if (task_delay_ms > LVGL_PORT_TASK_MAX_DELAY_MS) {
            task_delay_ms = LVGL_PORT_TASK_MAX_DELAY_MS;
        } else if (task_delay_ms < LVGL_PORT_TASK_MIN_DELAY_MS) {
            task_delay_ms = LVGL_PORT_TASK_MIN_DELAY_MS;
        }
        vTaskDelay(pdMS_TO_TICKS(task_delay_ms));
    }
}

// Mutex is created like this:
lv_mutex_init(&lvgl_mutex);                         // Make lvgl thread safe

// This is how the task is created
BaseType_t ret = xTaskCreatePinnedToCore(ProcessDisplay, "lvgl", LVGL_PORT_TASK_STACK_SIZE, NULL,
                                                                               LVGL_PORT_TASK_PRIORITY, &lvgl_task_handle, core_id);
// where stack size is 6k bytes, the piority is 2

Thanks for your suggestions, I tried them out with these results:

Adding the 2 lines with USE_FREERTOS_TASK_NOTIFY 0 --> many build issues, FreeRTOS does not compile successfully (see below some build output).

Adding the 2 lines with USE_FREERTOS_TASK_NOTIFY 1 --> exact same behavior as without these 2 lines, the system hangs.

So unfortunately this does not solve the issue, do you have other ideas I could try?


----------- Example of build errors ----------

Compiling .pio/build/esp32-s3-devkitm-1/lib158/lvgl/osal/lv_freertos.c.o
In file included from .pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.h:29,
                 from .pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_os.h:30,
                 from .pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.c:15:
.pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.c: In function 'lv_thread_sync_wait':
.pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.c:211:26: error: 'lv_thread_sync_t' {aka 'struct <anonymous>'} has no member named 'xSyncMutex'
     xSemaphoreTake(pxCond->xSyncMutex, portMAX_DELAY);
                          ^~
/home/john/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/include/freertos/include/freertos/semphr.h:312:76: note: in definition of macro 'xSemaphoreTake'
 #define xSemaphoreTake( xSemaphore, xBlockTime )    xQueueSemaphoreTake( ( xSemaphore ), ( xBlockTime ) )
                                                                            ^~~~~~~~~~
.pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.c:213:18: error: 'lv_thread_sync_t' {aka 'struct <anonymous>'} has no member named 'xSyncSignal'
     while(!pxCond->xSyncSignal) {
                  ^~
.pio/libdeps/esp32-s3-devkitm-1/lvgl/src/osal/lv_freertos.c:219:61: error: 'lv_thread_sync_t' {aka 'struct <anonymous>'} has no member named 'ulWaitingThreads'
         ulLocalWaitingThreads = Atomic_Increment_u32(&pxCond->ulWaitingThreads);

--- a/src/osal/lv_freertos.c
+++ b/src/osal/lv_freertos.c
@@ -179,8 +179,8 @@ lv_result_t lv_mutex_delete(lv_mutex_t * pxMutex)
 lv_result_t lv_thread_sync_init(lv_thread_sync_t * pxCond)
 {
 #if USE_FREERTOS_TASK_NOTIFY
-    /* Store the handle of the calling task. */
-    pxCond->xTaskToNotify = xTaskGetCurrentTaskHandle();
+    pxCond->xTaskToNotify = NULL;
+    pxCond->xSignalSent = pdFALSE;
 #else
     /* If the cond is uninitialized, perform initialization. */
     prvCheckCondInit(pxCond);
@@ -194,10 +194,21 @@ lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
     lv_result_t lvRes = LV_RESULT_OK;
 
 #if USE_FREERTOS_TASK_NOTIFY
-    LV_UNUSED(pxCond);
+    TaskHandle_t current_task_handle = xTaskGetCurrentTaskHandle();
 
-    /* Wait for other task to notify this task. */
-    ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+    _enter_critical();
+    BaseType_t signal_sent = pxCond->xSignalSent;
+    pxCond->xSignalSent = pdFALSE;
+    if(signal_sent == pdFALSE) {
+        /* The signal hasn't been sent yet. Tell the sender to notify this task */
+        pxCond->xTaskToNotify = current_task_handle;
+    }
+    _exit_critical();
+
+    if(signal_sent == pdFALSE) {
+        /* Wait for other task to notify this task. */
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+    }
 #else
     uint32_t ulLocalWaitingThreads;
 
@@ -261,8 +272,19 @@ lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
 lv_result_t lv_thread_sync_signal(lv_thread_sync_t * pxCond)
 {
 #if USE_FREERTOS_TASK_NOTIFY
-    /* Send a notification to the task waiting. */
-    xTaskNotifyGive(pxCond->xTaskToNotify);
+    _enter_critical();
+    TaskHandle_t task_to_notify = pxCond->xTaskToNotify;
+    pxCond->xTaskToNotify = NULL;
+    if(task_to_notify == NULL) {
+        /* No task waiting to be notified. Send this signal for later */
+        pxCond->xSignalSent = pdTRUE;
+    }
+    _exit_critical();
+
+    if(task_to_notify != NULL) {
+        /* There is a task waiting. Send a notification to it */
+        xTaskNotifyGive(task_to_notify);
+    }
 #else
     /* If the cond is uninitialized, perform initialization. */
     prvCheckCondInit(pxCond);
@@ -322,9 +344,20 @@ lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * pxCond)
 {
     BaseType_t xHigherPriorityTaskWoken = pdFALSE;
 #if USE_FREERTOS_TASK_NOTIFY
-    /* Send a notification to the task waiting. */
-    vTaskNotifyGiveFromISR(pxCond->xTaskToNotify, &xHigherPriorityTaskWoken);
-    portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+    _enter_critical();
+    TaskHandle_t task_to_notify = pxCond->xTaskToNotify;
+    pxCond->xTaskToNotify = NULL;
+    if(task_to_notify == NULL) {
+        /* No task waiting to be notified. Send this signal for later */
+        pxCond->xSignalSent = pdTRUE;
+    }
+    _exit_critical();
+
+    if(task_to_notify != NULL) {
+        /* There is a task waiting. Send a notification to it */
+        vTaskNotifyGiveFromISR(task_to_notify, &xHigherPriorityTaskWoken);
+        portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+    }
 #else
     /* If the cond is uninitialized, perform initialization. */
     prvCheckCondInit(pxCond);

--- a/src/osal/lv_freertos.c
+++ b/src/osal/lv_freertos.c
@@ -46,11 +46,11 @@ static void prvMutexInit(lv_mutex_t * pxMutex);
 
 static void prvCheckMutexInit(lv_mutex_t * pxMutex);
 
-#if !USE_FREERTOS_TASK_NOTIFY
 static void prvCondInit(lv_thread_sync_t * pxCond);
 
 static void prvCheckCondInit(lv_thread_sync_t * pxCond);
 
+#if !USE_FREERTOS_TASK_NOTIFY
 static void prvTestAndDecrement(lv_thread_sync_t * pxCond,
                                 uint32_t ulLocalWaitingThreads);
 #endif
@@ -83,7 +83,7 @@ lv_result_t lv_thread_init(lv_thread_t * pxThread, lv_thread_prio_t xSchedPriori
                            void (*pvStartRoutine)(void *), size_t usStackSize,
                            void * xAttr)
 {
-    pxThread->xTaskArg = xAttr;
+    pxThread->pTaskArg = xAttr;
     pxThread->pvStartRoutine = pvStartRoutine;
 
     BaseType_t xTaskCreateStatus = xTaskCreate(
@@ -178,13 +178,8 @@ lv_result_t lv_mutex_delete(lv_mutex_t * pxMutex)
 
 lv_result_t lv_thread_sync_init(lv_thread_sync_t * pxCond)
 {
-#if USE_FREERTOS_TASK_NOTIFY
-    pxCond->xTaskToNotify = NULL;
-    pxCond->xSignalSent = pdFALSE;
-#else
     /* If the cond is uninitialized, perform initialization. */
     prvCheckCondInit(pxCond);
-#endif
 
     return LV_RESULT_OK;
 }
@@ -192,6 +187,9 @@ lv_result_t lv_thread_sync_init(lv_thread_sync_t * pxCond)
 lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
 {
     lv_result_t lvRes = LV_RESULT_OK;
+
+    /* If the cond is uninitialized, perform initialization. */
+    prvCheckCondInit(pxCond);
 
 #if USE_FREERTOS_TASK_NOTIFY
     TaskHandle_t current_task_handle = xTaskGetCurrentTaskHandle();
@@ -211,9 +209,6 @@ lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
     }
 #else
     uint32_t ulLocalWaitingThreads;
-
-    /* If the cond is uninitialized, perform initialization. */
-    prvCheckCondInit(pxCond);
 
     /* Acquire the mutex. */
     xSemaphoreTake(pxCond->xSyncMutex, portMAX_DELAY);
@@ -271,6 +266,9 @@ lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
 
 lv_result_t lv_thread_sync_signal(lv_thread_sync_t * pxCond)
 {
+    /* If the cond is uninitialized, perform initialization. */
+    prvCheckCondInit(pxCond);
+
 #if USE_FREERTOS_TASK_NOTIFY
     _enter_critical();
     TaskHandle_t task_to_notify = pxCond->xTaskToNotify;
@@ -286,9 +284,6 @@ lv_result_t lv_thread_sync_signal(lv_thread_sync_t * pxCond)
         xTaskNotifyGive(task_to_notify);
     }
 #else
-    /* If the cond is uninitialized, perform initialization. */
-    prvCheckCondInit(pxCond);
-
     /* Acquire the mutex. */
     xSemaphoreTake(pxCond->xSyncMutex, portMAX_DELAY);
 
@@ -326,16 +321,14 @@ lv_result_t lv_thread_sync_signal(lv_thread_sync_t * pxCond)
 
 lv_result_t lv_thread_sync_delete(lv_thread_sync_t * pxCond)
 {
-#if USE_FREERTOS_TASK_NOTIFY
-    LV_UNUSED(pxCond);
-#else
+#if !USE_FREERTOS_TASK_NOTIFY
     /* Cleanup all resources used by the cond. */
     vSemaphoreDelete(pxCond->xCondWaitSemaphore);
     vSemaphoreDelete(pxCond->xSyncMutex);
     pxCond->ulWaitingThreads = 0;
     pxCond->xSyncSignal = pdFALSE;
-    pxCond->xIsInitialized = pdFALSE;
 #endif
+    pxCond->xIsInitialized = pdFALSE;
 
     return LV_RESULT_OK;
 }
@@ -343,6 +336,10 @@ lv_result_t lv_thread_sync_delete(lv_thread_sync_t * pxCond)
 lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * pxCond)
 {
     BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+    /* If the cond is uninitialized, perform initialization. */
+    prvCheckCondInit(pxCond);
+
 #if USE_FREERTOS_TASK_NOTIFY
     _enter_critical();
     TaskHandle_t task_to_notify = pxCond->xTaskToNotify;
@@ -359,9 +356,6 @@ lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * pxCond)
         portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
     }
 #else
-    /* If the cond is uninitialized, perform initialization. */
-    prvCheckCondInit(pxCond);
-
     /* Enter critical section to prevent preemption. */
     _enter_critical();
 
@@ -390,7 +384,7 @@ static void prvRunThread(void * pxArg)
     lv_thread_t * pxThread = (lv_thread_t *)pxArg;
 
     /* Run the thread routine. */
-    pxThread->pvStartRoutine((void *)pxThread->xTaskArg);
+    pxThread->pvStartRoutine((void *)pxThread->pTaskArg);
 
     vTaskDelete(NULL);
 }
@@ -429,9 +423,12 @@ static void prvCheckMutexInit(lv_mutex_t * pxMutex)
     }
 }
 
-#if !USE_FREERTOS_TASK_NOTIFY
 static void prvCondInit(lv_thread_sync_t * pxCond)
 {
+#if USE_FREERTOS_TASK_NOTIFY
+    pxCond->xTaskToNotify = NULL;
+    pxCond->xSignalSent = pdFALSE;
+#else
     pxCond->xCondWaitSemaphore = xSemaphoreCreateCounting(ulMAX_COUNT, 0U);
 
     /* Ensure that the FreeRTOS semaphore was successfully created. */
@@ -453,13 +450,13 @@ static void prvCondInit(lv_thread_sync_t * pxCond)
     /* Condition variable successfully created. */
     pxCond->ulWaitingThreads = 0;
     pxCond->xSyncSignal = pdFALSE;
+#endif
+
     pxCond->xIsInitialized = pdTRUE;
 }
 
 static void prvCheckCondInit(lv_thread_sync_t * pxCond)
 {
-    BaseType_t xSemCreateStatus = pdTRUE;
-
     /* Check if the condition variable needs to be initialized. */
     if(pxCond->xIsInitialized == pdFALSE) {
         /* Cond initialization must be in a critical section to prevent two
@@ -478,6 +475,7 @@ static void prvCheckCondInit(lv_thread_sync_t * pxCond)
     }
 }
 
+#if !USE_FREERTOS_TASK_NOTIFY
 static void prvTestAndDecrement(lv_thread_sync_t * pxCond,
                                 uint32_t ulLocalWaitingThreads)
 {

--- a/src/osal/lv_freertos.c
+++ b/src/osal/lv_freertos.c
@@ -195,18 +195,20 @@ lv_result_t lv_thread_sync_wait(lv_thread_sync_t * pxCond)
     TaskHandle_t current_task_handle = xTaskGetCurrentTaskHandle();
 
     _enter_critical();
-    BaseType_t signal_sent = pxCond->xSignalSent;
-    pxCond->xSignalSent = pdFALSE;
+    BaseType_t signal_sent = pxCond->xSyncSignal;
+    pxCond->xSyncSignal = pdFALSE;
     if(signal_sent == pdFALSE) {
         /* The signal hasn't been sent yet. Tell the sender to notify this task */
         pxCond->xTaskToNotify = current_task_handle;
     }
+    /* If we have a signal from the other task, we should not ask to be notified */
     _exit_critical();
 
     if(signal_sent == pdFALSE) {
         /* Wait for other task to notify this task. */
         ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
     }
+    /* If the signal was received, no wait needs to be done */
 #else
     uint32_t ulLocalWaitingThreads;
 
@@ -275,14 +277,16 @@ lv_result_t lv_thread_sync_signal(lv_thread_sync_t * pxCond)
     pxCond->xTaskToNotify = NULL;
     if(task_to_notify == NULL) {
         /* No task waiting to be notified. Send this signal for later */
-        pxCond->xSignalSent = pdTRUE;
+        pxCond->xSyncSignal = pdTRUE;
     }
+    /* If a task is already waiting, there is no need to set the sync signal */
     _exit_critical();
 
     if(task_to_notify != NULL) {
         /* There is a task waiting. Send a notification to it */
         xTaskNotifyGive(task_to_notify);
     }
+    /* If there was no task waiting to be notified, we sent a signal for it to see later. */
 #else
     /* Acquire the mutex. */
     xSemaphoreTake(pxCond->xSyncMutex, portMAX_DELAY);
@@ -346,8 +350,9 @@ lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * pxCond)
     pxCond->xTaskToNotify = NULL;
     if(task_to_notify == NULL) {
         /* No task waiting to be notified. Send this signal for later */
-        pxCond->xSignalSent = pdTRUE;
+        pxCond->xSyncSignal = pdTRUE;
     }
+    /* If a task is already waiting, there is no need to set the sync signal */
     _exit_critical();
 
     if(task_to_notify != NULL) {
@@ -355,6 +360,7 @@ lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * pxCond)
         vTaskNotifyGiveFromISR(task_to_notify, &xHigherPriorityTaskWoken);
         portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
     }
+    /* If there was no task waiting to be notified, we sent a signal for it to see later. */
 #else
     /* Enter critical section to prevent preemption. */
     _enter_critical();
@@ -425,9 +431,11 @@ static void prvCheckMutexInit(lv_mutex_t * pxMutex)
 
 static void prvCondInit(lv_thread_sync_t * pxCond)
 {
+    pxCond->xIsInitialized = pdTRUE;
+    pxCond->xSyncSignal = pdFALSE;
+
 #if USE_FREERTOS_TASK_NOTIFY
     pxCond->xTaskToNotify = NULL;
-    pxCond->xSignalSent = pdFALSE;
 #else
     pxCond->xCondWaitSemaphore = xSemaphoreCreateCounting(ulMAX_COUNT, 0U);
 
@@ -449,10 +457,7 @@ static void prvCondInit(lv_thread_sync_t * pxCond)
 
     /* Condition variable successfully created. */
     pxCond->ulWaitingThreads = 0;
-    pxCond->xSyncSignal = pdFALSE;
 #endif
-
-    pxCond->xIsInitialized = pdTRUE;
 }
 
 static void prvCheckCondInit(lv_thread_sync_t * pxCond)

--- a/src/osal/lv_freertos.h
+++ b/src/osal/lv_freertos.h
@@ -51,7 +51,7 @@ extern "C" {
 
 typedef struct {
     void (*pvStartRoutine)(void *);       /**< Application thread function. */
-    void * xTaskArg;                      /**< Arguments for application thread function. */
+    void * pTaskArg;                      /**< Arguments for application thread function. */
     TaskHandle_t xTaskHandle;             /**< FreeRTOS task handle. */
 } lv_thread_t;
 
@@ -61,12 +61,12 @@ typedef struct {
 } lv_mutex_t;
 
 typedef struct {
+    BaseType_t
+    xIsInitialized;                       /**< Set to pdTRUE if this condition variable is initialized, pdFALSE otherwise. */
 #if USE_FREERTOS_TASK_NOTIFY
     TaskHandle_t xTaskToNotify;
     BaseType_t xSignalSent;
 #else
-    BaseType_t
-    xIsInitialized;                       /**< Set to pdTRUE if this condition variable is initialized, pdFALSE otherwise. */
     SemaphoreHandle_t xCondWaitSemaphore; /**< Threads block on this semaphore in lv_thread_sync_wait. */
     uint32_t ulWaitingThreads;            /**< The number of threads currently waiting on this condition variable. */
     SemaphoreHandle_t xSyncMutex;         /**< Threads take this mutex before accessing the condition variable. */

--- a/src/osal/lv_freertos.h
+++ b/src/osal/lv_freertos.h
@@ -63,14 +63,13 @@ typedef struct {
 typedef struct {
     BaseType_t
     xIsInitialized;                       /**< Set to pdTRUE if this condition variable is initialized, pdFALSE otherwise. */
+    BaseType_t xSyncSignal;               /**< Set to pdTRUE if the thread is signaled, pdFALSE otherwise. */
 #if USE_FREERTOS_TASK_NOTIFY
-    TaskHandle_t xTaskToNotify;
-    BaseType_t xSignalSent;
+    TaskHandle_t xTaskToNotify;           /**< The task waiting to be signalled. NULL if nothing is waiting. */
 #else
     SemaphoreHandle_t xCondWaitSemaphore; /**< Threads block on this semaphore in lv_thread_sync_wait. */
     uint32_t ulWaitingThreads;            /**< The number of threads currently waiting on this condition variable. */
     SemaphoreHandle_t xSyncMutex;         /**< Threads take this mutex before accessing the condition variable. */
-    BaseType_t xSyncSignal;               /**< Set to pdTRUE if the thread is signaled, pdFALSE otherwise. */
 #endif
 } lv_thread_sync_t;
 

--- a/src/osal/lv_freertos.h
+++ b/src/osal/lv_freertos.h
@@ -63,6 +63,7 @@ typedef struct {
 typedef struct {
 #if USE_FREERTOS_TASK_NOTIFY
     TaskHandle_t xTaskToNotify;
+    BaseType_t xSignalSent;
 #else
     BaseType_t
     xIsInitialized;                       /**< Set to pdTRUE if this condition variable is initialized, pdFALSE otherwise. */

